src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/custom.py

```diff
@@ -0,0 +1,176 @@
+        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as err:
+            raise CLIError('Please ensure you have network connection. Error detail: {}'.format(str(err)))
+        logger.debug('Downloaded to %s', ext_file)
+    else:
+        # Get file path
+        ext_file = os.path.realpath(os.path.expanduser(source))
+        if not os.path.isfile(ext_file):
+            raise CLIError("File {} not found.".format(source))
+    # Validate the extension
+    logger.debug('Validating the extension {}'.format(ext_file))
+    try:
+        _validate_whl_extension(ext_file)
+    except AssertionError:
+        logger.debug(traceback.format_exc())
+        raise CLIError('The extension is invalid. Use --debug for more information.')
+    except CLIError as e:
+        raise e
+    except Exception:  # pylint:disable=broad-except
+        logger.debug(traceback.format_exc())
+        raise CLIError('Error validating extension. Use --debug for more information.')
+    logger.debug('Validation successful on {}'.format(ext_file))
+    # Install with pip
+    extension_path = get_extension_path(extension_name)
+    pip_args = ['install', '--target', extension_path, ext_file]
+    logger.debug('Executing pip with args: %s', pip_args)
+    pip_status_code = _run_pip(pip, pip_args)
+    if pip_status_code > 0:
+        logger.debug('Pip failed so deleting anything we might have installed at %s', extension_path)
+        shutil.rmtree(extension_path, ignore_errors=True)
+        raise CLIError('An error occurred. Pip failed with status code {}. '
+                       'Use --debug for more information.'.format(pip_status_code))
+    # Save the whl we used to install the extension in the extension dir.
+    dst = os.path.join(extension_path, whl_filename)
+    shutil.copyfile(ext_file, dst)
+    logger.debug('Saved the whl to %s', dst)
+
+
+def add_extension(source):
+    if source.endswith('.whl'):
+        _add_whl_ext(source)
+    else:
+        raise CLIError('Unknown extension type. Only Python wheels are supported.')
+
+
+def remove_extension(extension_name):
+    try:
+        get_extension(extension_name)
+        shutil.rmtree(get_extension_path(extension_name))
+    except ExtensionNotInstalledException as e:
+        raise CLIError(e)
+
+
+def list_extensions():
+    extensions = []
+    for ext in get_extensions():
+        extensions.append({OUT_KEY_NAME: ext.name,
+                           OUT_KEY_VERSION: ext.version,
+                           OUT_KEY_TYPE: ext.ext_type})
+    return extensions
+
+
+def show_extension(extension_name):
+    try:
+        extension = get_extension(extension_name)
+        return {OUT_KEY_NAME: extension.name,
+                OUT_KEY_VERSION: extension.version,
+                OUT_KEY_TYPE: extension.ext_type,
+                OUT_KEY_METADATA: extension.metadata}
+    except ExtensionNotInstalledException as e:
+        raise CLIError(e)
```

Comment:
I didn't want to use `CLIError` inside the `extension.py` file so `list_extensions` will raise an `ExtensionNotInstalledException` exception then the caller can determine how to handle it.